src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/PkgControl.java

```diff
@@ -40,28 +47,130 @@
     private final PkgControl parent;
     /** The full package name for the node. */
     private final String fullPackage;
+    /**
+     * The regex pattern for partial match (exact and for subpackages) - only not
+     * null if regex is true.
+     */
+    private final Pattern patternForPartialMatch;
+    /** The regex pattern for exact matches - only not null if regex is true. */
+    private final Pattern patternForExactMatch;
+    /** If this package represents a regular expression. */
+    private final boolean regex;
 
     /**
      * Construct a root node.
      * @param pkgName the name of the package.
+     * @param regex flags interpretation of pkgName as regex pattern.
      */
-    PkgControl(final String pkgName) {
+    PkgControl(final String pkgName, final boolean regex) {
         parent = null;
-        fullPackage = pkgName;
+        this.regex = regex;
+        fullPackage = addGroupingIfNecessary(pkgName, regex);
+        if (regex) {
+            patternForPartialMatch = createPatternForPartialMatch();
+            patternForExactMatch = createPatternForExactMatch();
+        }
+        else {
+            patternForPartialMatch = null;
+            patternForExactMatch = null;
+        }
     }
 
     /**
      * Construct a child node.
      * @param parent the parent node.
      * @param subPkg the sub package name.
+     * @param regex flags interpretation of subPkg as regex pattern.
      */
-    PkgControl(final PkgControl parent, final String subPkg) {
+    PkgControl(final PkgControl parent, final String subPkg, final boolean regex) {
         this.parent = parent;
-        fullPackage = parent.fullPackage + "." + subPkg;
+        if (regex || parent.regex) {
+            // regex gets inherited
+            final String parentRegex = ensureRegex(parent.fullPackage, parent.regex);
```

Comment:
we need to add groups only at this line , when we do inheritance of regexp. In all cases it is not required.
Comment above need to be updated to attract user attention that only here we need grouping.